### DISCUSSION TITLE
Normalization of hybrid retrieval

### DISCUSSION THREAD

**Discussion by vrdn-23** [castorini/pyserini#1413] [Q&A]
Is there any literature that points to the normalization implementation for scores in Hybrid retrieval? Is there any reason why we normalize such that certain documents may end up with a negative score, instead of say normalizing both (sparse and dense) scores to between 0 and 1?

**Comment by GioiaZheng** [comment]
I would separate two cases here.

At least in the current `pyserini/search/hybrid/_searcher.py`, `HybridSearcher(normalization=True)` is using per-query range centering rather than ordinary min-max normalization:

`s' = (s - (min + max) / 2) / (max - min)`

So, assuming the run has a non-constant score range, the minimum observed score maps to `-0.5` and the maximum maps to `0.5`. Documents that appear only in the sparse or dense list are assigned the other side's minimum score before the interpolation step, so a negative contribution is expected. The final score is mainly a ranking device; it does not have an absolute probability-like meaning.

The practical reason is that BM25 and dense scores are not on the same scale. Centering gives each side both positive and negative evidence around that query's own midpoint before applying `alpha`. Mapping both sides to `[0, 1]` is also a reasonable choice, but it changes the behavior: every retrieved document contributes non-negative evidence, and documents missing from one list get the floor instead of a negative/lowest signal. I would normally tune the normalizer and `alpha` on validation queries rather than treat either transform as universally correct.

For literature, this is related to the classic data-fusion / metasearch score-normalization line: Fox and Shaw's TREC fusion methods such as CombSUM/CombMNZ, Lee's analysis of multiple evidence combination, and Montague and Aslam's work on relevance-score normalization for metasearch. If the score scales are hard to trust, Reciprocal Rank Fusion is often a safer baseline because it avoids direct score calibration.

One caveat: the separate `pyserini.fusion.normalize` command path is different. It min-max normalizes each input run to `[0, 1]` before summing, so if you specifically want `[0, 1]` run normalization, that API path is closer to what you described.
